File: packages/guards/src/json-depth-limit.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface JsonDepthLimitOptions {
  action: 'block' | 'warn';
  maxDepth?: number;
}

function measureDepth(text: string): number {
  let maxDepth = 0;
  let current = 0;
  for (const ch of text) {
    if (ch === '{' || ch === '[') {
      current++;
      if (current > maxDepth) maxDepth = current;
    } else if (ch === '}' || ch === ']') {
      current--;
    }
  }
  return maxDepth;
}

export function jsonDepthLimit(options: JsonDepthLimitOptions): Guard {
  const maxDepth = options.maxDepth ?? 20;

  return {
    name: 'json-depth-limit',
    version: '0.1.0',
    description: 'Limits JSON nesting depth to prevent DoS attacks',
    category: 'format',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const depth = measureDepth(text);
      const triggered = depth > maxDepth;

      return {
        guardName: 'json-depth-limit',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(depth / (maxDepth * 2), 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: { depth, maxDepth },
      };
    },
  };
}

File: python/open_guardrail/guards/json_depth_limit.py
"""Limit JSON nesting depth to prevent DoS."""
from __future__ import annotations

import time

from open_guardrail.core import GuardResult


def _measure_depth(text: str) -> int:
    max_d = 0
    cur = 0
    for ch in text:
        if ch in ("{", "["):
            cur += 1
            if cur > max_d:
                max_d = cur
        elif ch in ("}", "]"):
            cur -= 1
    return max_d


class _JsonDepthLimit:
    def __init__(
        self,
        *,
        action: str = "block",
        max_depth: int = 20,
    ) -> None:
        self.name = "json-depth-limit"
        self.action = action
        self._max = max_depth

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        depth = _measure_depth(text)
        triggered = depth > self._max
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="json-depth-limit",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(depth / (self._max * 2), 1.0)
                if triggered
                else 0.0
            ),
            message=(
                f"JSON depth {depth} exceeds {self._max}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "depth": depth,
                "max_depth": self._max,
            },
        )


def json_depth_limit(
    *,
    action: str = "block",
    max_depth: int = 20,
) -> _JsonDepthLimit:
    return _JsonDepthLimit(
        action=action, max_depth=max_depth
    )
